core/src/main/scala/org/scalafmt/internal/Router.scala

```diff
@@ -420,67 +420,114 @@ class Router(formatOps: FormatOps) {
         }
 
       // Parameter opening for one parameter group. This format works
-      // on a group-by-group basis.
-      case FormatToken(open @ LeftParen(), r, between)
+      // on the WHOLE defnSite (via policies)
+      case FormatToken(open @ (LeftParen() | LeftBracket()), r, between)
           if style.verticalMultilineAtDefinitionSite &&
-            isDefnSite(leftOwner) =>
+            isClassOrDef(leftOwner) =>
+
         val close = matchingParentheses(hash(open))
         val indentParam = Num(style.continuationIndent.defnSite)
         val indentSep = Num((indentParam.n - 2).max(0))
+        val isBracket = open.is[LeftBracket]
 
         @tailrec
-        def loop(token: Token): Option[Token] =
+        def loop(token: Token): Option[Token] = {
           leftTok2tok(matchingParentheses(hash(token))) match {
-            case FormatToken(RightParen(), l @ LeftParen(), _) => loop(l)
-            case FormatToken(r @ RightParen(), _, _) => Some(r)
+            case FormatToken(RightParenOrBracket(), l @ LeftParen(), _) =>
+              loop(l)
+            // This case only applies to classes
+            case f @ FormatToken(RightBracket(), CtorModifier(), _) =>
+              loop(next(f).right)
+            case FormatToken(r @ RightParenOrBracket(), _, _) => Some(r)
             case _ => None
           }
-        // find the last param on the def
-        // so that we can apply our `policy` till
-        // the end.
+        }
+        // find the last param on the defn so that we can apply our `policy`
+        // till the end.
         val lastParen = loop(open).get
 
-        // If this is a class, we need to do some different things
+        // If this is a class/trait, we need to do some things differently.
         // We only care about the Primary Ctor since other Ctors will
         // be picked up by the `def` definition
-        val isCtor = leftOwner.is[meta.Ctor.Primary]
+        val isClassTrait =
+          leftOwner.is[meta.Ctor.Primary] ||
+          leftOwner.is[meta.Defn.Class] ||
+          leftOwner.is[meta.Defn.Trait]
```

Comment:
Since I have the holistic information about the defn (ie: I can ask things like "does this have any data params? Any type params?") I feel like this can all be cleaner (at the cost of, maybe, some code duplication).